### PR TITLE
Minor: Move function signature check to planning stage

### PR DESCRIPTION
**Pull request by 2010YOUY01**
## Which issue does this PR close?

NA

## Rationale for this change

I noticed a recent PR https://github.com/apache/arrow-datafusion/pull/9377 tried to check function argument number inside a specific function's implementation
It can be done at a higher level during planning (as it was in old builtin function implementation)

## What changes are included in this PR?

1. Move checking function argument number logic outside of function implementation body (for all recently ported function did such check)
2. Check argument number during planning (same as old builtin function's implementation)

## Are these changes tested?
Covered by existing test introduced in #9377 

## Are there any user-facing changes?
No

### CODE REVIEW

File: datafusion/expr/src/expr_schema.rs [hunk 1]
```diff
@@ -134,9 +134,25 @@ impl ExprSchemable for Expr {
...
+                        data_types(&arg_data_types, fun.signature()).map_err(|_| {
+                            plan_datafusion_err!(
+                                "{}",
+                                utils::generate_signature_error_msg(
+                                    fun.name(),
+                                    fun.signature().clone(),
+                                    &arg_data_types,
+                                )
+                            )
+                        })?;
```
Comment (SteveLauC) [inline datafusion/expr/src/expr_schema.rs:152]: Thanks for this! Doing it during the planning stage eliminates so much duplicate code:)

FYI @jonahgao, this addresses the TODO we talked about in this [comment](https://github.com/apache/arrow-datafusion/pull/9297#issuecomment-1958566729)

Comment (jonahgao) [reply datafusion/expr/src/expr_schema.rs:152]: Nice! 👍

**Review by comphead — APPROVED** [review]
lgtm thanks @2010YOUY01 
Thats good start to unify error messages if signature cannot be found

File: datafusion/sqllogictest/test_files/errors.slt [hunk 2]
```diff
@@ -84,7 +84,7 @@ statement error Error during planning: No function matches the given name and ar
 SELECT concat();
 
 # error message for wrong function signature (Uniform: t args all from some common types)
-statement error DataFusion error: Failed to coerce arguments for NULLIF
+statement error DataFusion error: Error during planning: No function matches the given name and argument types 'nullif\(Int64\)'. You might need to add explicit type casts.
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/errors.slt:87]: That is a much nicer error message

**Review by alamb — APPROVED** [review]
Thank you @2010YOUY01  (and welcome back!)

**Comment by jayzhan211** [comment]
Let me merge this nice change

**Pull request merged**